File: vob/demo/between.py
# 
# Copyright (c) 2005, Matti J. Katila and Benja Fallenstain
# This file is part of Libvob.
# 
# Libvob is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# Libvob is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
# Public License for more details.
# 
# You should have received a copy of the GNU General
# Public License along with Libvob; if not, write to the Free
# Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
# MA  02111-1307  USA
# 


import vob


class Scene:
    def __init__(self):
        self.mousePt = [300, 300]

    def mouse(self, ev):
        self.mousePt = [ev.getX(), ev.getY()]
        if hasattr(self, 'vs'):
            self.vs.anim.switchVS()

    def scene(self, vs):
        self.vs = vs
        vs.put(background((.1, .1, .9)))

        a = vs.coords.translate(0, 100, 100, 0)
        a = vs.coords.scale(a, 10, 10)
        a = vs.coords.box(a, 30, 5)

        b = vs.coords.translate(0, self.mousePt[0], self.mousePt[1], 0)
        b = vs.coords.scale(b, 10, 10)
        b = vs.coords.box(b, 30, 5)

        ac = vs.coords.center(a)
        bc = vs.coords.center(b)

        ac = vs.coords.box(ac, 0, 0)
        bc = vs.coords.box(bc, 0, 0)

        ac = vs.coords.translate(ac, 0, 0, .5)
        bc = vs.coords.translate(bc, 0, 0, .5)

        vs.put(vob.vobs.SimpleConnection(.5,.5,.5,.5, java.awt.Color.white),
               ac, bc)

        c = vs.coords.between(ac, bc)
        vs.put(vob.vobs.RectBgVob(), a)
        vs.put(vob.vobs.RectBgVob(), b)
        cs = vs.orthoBoxCS(c, "buoy", 0, -10,-10, 1,1, 20,20)
        vs.put(vob.vobs.OvalBgVob(java.awt.Color.yellow),
               cs)
        
        

File: org/nongnu/libvob/lob/lobs/Between.java
/*
Between.java
 *    
 *    Copyright (c) 2004-2005, Benja Fallenstein
 *
 *    This file is part of libvob.
 *    
 *    libvob is free software; you can redistribute it and/or modify it under
 *    the terms of the GNU General Public License as published by
 *    the Free Software Foundation; either version 2 of the License, or
 *    (at your option) any later version.
 *    
 *    libvob is distributed in the hope that it will be useful, but WITHOUT
 *    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 *    or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 *    Public License for more details.
 *    
 *    You should have received a copy of the GNU General
 *    Public License along with libvob; if not, write to the Free
 *    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
 *    MA  02111-1307  USA
 *    
 *
 */
/*
 * Written by Benja Fallenstein
 */
package org.nongnu.libvob.lob.lobs;
import org.nongnu.libvob.lob.*;
import org.nongnu.libvob.*;
import java.awt.Color;
import java.util.*;

/** A lob drawing a lobs in front of and behind another lob.
 */
public class Between extends AbstractLob {

    protected Lob back, middle, front;

    private Between() {}

    public Between newInstance(Lob back, Lob middle, Lob front) {
	Between b = (Between)LOB_FACTORY.object();
	
	b.back   = (back == null) ?   NullLob.instance : back;
	b.middle = (middle == null) ? NullLob.instance : middle;
	b.front  = (front == null) ?  NullLob.instance : front;

	return b;
    }

    public SizeRequest getSizeRequest() {
	return middle.getSizeRequest();
    }

    public Layout layout(float w, float h) {
	BetweenLayout l = (BetweenLayout)LAYOUT_FACTORY.object();
	l.back   = back.layout(w, h);
	l.middle = middle.layout(w, h);
	l.front  = front.layout(w, h);
	return l;
    }

    public boolean move(ObjectSpace os) {
	if(super.move(os)) {
	    back.move(os); middle.move(os); front.move(os);
	    return true;
	}
	return false;
    }

    protected static class BetweenLayout extends AbstractLayout {
	private Layout back, middle, front;

	public Size getSize() {
	    return middle.getSize();
	}

	public void render(VobScene scene, int into, int matchingParent,
			   float d, boolean visible) {
	    int _cs = into;
	    int cs = scene.coords.translate(_cs, 0, 0, 2*d/4);
	    back.render(scene, cs, matchingParent, d/4, visible);
	    cs = scene.coords.translate(_cs, 0, 0, d/4);
	    middle.render(scene, cs, matchingParent, d/4, visible);
	    cs = _cs;
	    front.render(scene, cs, matchingParent, d/4, visible);
	}

	public boolean move(ObjectSpace os) {
	    if(super.move(os)) {
		back.move(os); middle.move(os); front.move(os);
		return true;
	    }
	    return false;
	}
    }

    private static final Factory LAYOUT_FACTORY = new Factory() {
	    public Object create() {
		return new BetweenLayout();
	    }
	};

    private static final Factory LOB_FACTORY = new Factory() {
	    public Object create() {
		return new Between();
	    }
	};
}
